inst/include/Rcpp/String.h

```diff
@@ -329,11 +347,46 @@ namespace Rcpp {
             return std::wstring( s, s + strlen(s) );
         }
 
-
         inline const char* get_cstring() const {
             return buffer_ready ? buffer.c_str() : CHAR(data) ;
         }
 
+        inline const std::string get_encoding() const {
+            switch (enc) {
+                case CE_BYTES:
+                    return "bytes";
+                case CE_LATIN1:
+                    return "latin1";
+                case CE_UTF8:
+                    return "UTF-8";
+                default:
+                    return "unknown";
+            }
+        }
+
+        inline void set_encoding( cetype_t encoding ) {
+            enc = encoding;
+
+            if (valid) {
+                data = Rf_mkCharCE(Rf_translateCharUTF8(data), encoding);
```

Comment:
From Writing R Extensions (
http://cran.rstudio.com/doc/manuals/r-devel/R-exts.html#Handling-character-data
):

CHARSXPs are read-only objects and must never be modified. In particular,
the C-style string contained in a CHARSXP should be treated as read-only
and for this reason the CHAR function used to access the character data of
a CHARSXP returns (const char *) (this also allows compilers to issue
warnings about improper use). Since CHARSXPs are immutable, the same
CHARSXP can
be shared by any STRSXP needing an element representing the same string. R
maintains a global cache of CHARSXPs so that there is only ever one
CHARSXP representing
a given string in memory.


On Mon, Jul 6, 2015 at 1:25 PM, Kevin Ushey <notifications@github.com>
wrote:

> In inst/include/Rcpp/String.h
> <https://github.com/RcppCore/Rcpp/pull/310#discussion_r33959013>:
>
> > +                case CE_BYTES:
> > +                    return "bytes";
> > +                case CE_LATIN1:
> > +                    return "latin1";
> > +                case CE_UTF8:
> > +                    return "UTF-8";
> > +                default:
> > +                    return "unknown";
> > +            }
> > +        }
> > +
> > +        inline void set_encoding( cetype_t encoding ) {
> > +            enc = encoding;
> > +
> > +            if (valid) {
> > +                data = Rf_mkCharCE(Rf_translateCharUTF8(data), encoding);
>
> I think CHARSXPs enter a global string cache (it's something I vaguely
> remember reading about a long time ago) but I think it's worth confirming /
> testing. It should be easy to make an example with e.g. an unprotected
> Rf_mkChar + a following Rf_allocVector + gctorture; presumedly that
> CHARSXP would be cleaned up if the GC touched it.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/RcppCore/Rcpp/pull/310/files#r33959013>.
>